plugins/src/main/java/tech/pegasys/pantheon/plugins/internal/PantheonPluginContextImpl.java

```diff
@@ -0,0 +1,170 @@
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.plugins.internal;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+
+import tech.pegasys.pantheon.plugins.PantheonContext;
+import tech.pegasys.pantheon.plugins.PantheonPlugin;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import java.util.stream.Stream;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class PantheonPluginContextImpl implements PantheonContext {
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  private enum Lifecycle {
+    UNINITIALIZED,
+    REGISTERING,
+    REGISTERED,
+    STARTING,
+    STARTED,
+    STOPPING,
+    STOPPED
+  }
+
+  private Lifecycle state = Lifecycle.UNINITIALIZED;
+  private final Map<Class<?>, ? super Object> serviceRegistry = new HashMap<>();
+  private final List<PantheonPlugin> plugins = new ArrayList<>();
+
+  public <T> void addService(final Class<T> serviceType, final T service) {
+    checkArgument(serviceType.isInterface(), "Services must be Java interfaces.");
+    checkArgument(
+        serviceType.isInstance(service),
+        "The service registered with a type must implement that type");
+    serviceRegistry.put(serviceType, service);
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T> Optional<T> getService(final Class<T> serviceType) {
+    return Optional.ofNullable((T) serviceRegistry.get(serviceType));
+  }
+
+  public void registerPlugins(final Path pluginsDir) {
+    checkState(state == Lifecycle.UNINITIALIZED, "PantheonContext has already been registered.");
+    if (pluginsDir == null) {
+      LOG.debug("Plugins are disabled.");
+      return;
+    }
+    state = Lifecycle.REGISTERING;
+
+    LOG.debug("Searching for plugins in {}", pluginsDir.toAbsolutePath().toString());
+    try (final Stream<Path> pluginFilesList = Files.list(pluginsDir)) {
+      final URL[] pluginJarURLs =
+          pluginFilesList
+              .filter(p -> p.getFileName().toString().endsWith(".jar"))
+              .map(PantheonPluginContextImpl::pathToURIOrNull)
+              .toArray(URL[]::new);
+      final ServiceLoader<PantheonPlugin> serviceLoader =
+          ServiceLoader.load(
+              PantheonPlugin.class,
+              new URLClassLoader(pluginJarURLs, this.getClass().getClassLoader()));
+
+      for (final PantheonPlugin plugin : serviceLoader) {
+        try {
+          plugin.register(this);
```

Comment:
Is it a little dangerous to give plugins access to the object that can start / stop / register plugins?  Why not just have a separate object that holds your services?

Reply:
This is written with Java modules in mind.  The class they would need to cast it to (PantheonPluginContextImpl) will not be visible in the modules and the JVM will enforce that barrier at compile time and runtime.  The interface only exposes a PantheonContext to them.

If we used a proxy class they will still be able to (pre Java modules) use reflection to navigate to the containing context and get access.

Plugins right now are not meant to run byzantine plugins.

If I am missing something ping me on slack.